### DISCUSSION TITLE
Twirl templates no longer import TemplateN traits in Play 3.0.3 for Scala 3.

### DISCUSSION THREAD

**Discussion by colin-lamed** [playframework/playframework#12643] [Q&A]
Generated twirl templates no-longer import the `TemplateN` traits (which are still present in the jar) in Play 3.0.3 for Scala 3 (This [commit](https://github.com/playframework/twirl/commit/75eec7cf46e23f0d10fa0a7f24d2f9b44c6c9738))

Is there a particular reason for this? I appreciate that most twirl users will probably not be affected by this change, but we have tests that generically check that our twirl components render the expected html.

They look something like

```scala
abstract class Template1UnitSpec[T, C <: Template1[T, HtmlFormat.Appendable]] {

  def render(params: T): HtmlFormat.Appendable =
    app.injector.instanceOf[C].render(params)
}
```

Note, we have been unable to use structural typing to access the render function (which is still present)

```scala
import scala.reflect.Selectable.reflectiveSelectable
abstract class Template1UnitSpec[T, C <: { def render(t: T): HtmlFormat.Appendable }]
```

since this gives

> "Structural access not allowed on method render because it has a parameter type with an unstable erasure".

If these traits are no longer going to be implemented by the templates, are there any recommended alternatives?

**Comment by mkurz** [comment]
Thanks for reporting this, this is an interesting use case. (Good to see HMRC sticking to Play btw :+1:)

> Is there a particular reason for this?

For me `TemplateN` looked like a bit of technical debt that was added long long time ago when Scala 2.10 was still a thing. The only reason I could think of having `TemplateN` was to have nicer error messages, see

- https://github.com/scala/bug/issues/10737

starting with Scala 2.13 the error messages are much clear now, thanks to

- https://github.com/scala/scala/pull/6401

so you now end up with e.g.
```
[error] ./play-scala-seed/app/views/index.scala.html:6:2: function values may not have more than 22 parameters, but 26 given
```

so I actually thought to totally remove `TemplateN` from twirl since I couldn't see any real use for it, however it turned out that we still support Scala 2.12, so I kept it:
- https://github.com/playframework/twirl/blob/2.0.5/project/Dependencies.scala#L9
- https://github.com/playframework/twirl/blob/2.0.5/project/Common.scala#L53

(and of course the twirl sbt plugins is still Scala 2.12 because sbt 1.0 uses Scala 2.12)

(Hmm... thinking out loud, actually why didn't I drop it for Scala 2.13 then?)

Now... thinking about your use case, we can go the opposite direction and re-add `TemplateN` even to Scala 3 templates, but only up to 22 parameters. This would "fix" your use case and I guess would work for 99% lof the people. I mean I doubt that you ever will have templates exceeding 22 params (if you haven't had that until today)...

What do you think? I will come up with a pull request in the twirl repo and ping you from there. I could cut a twirl release quite fast so you can make use of it and not have to wait for the next Play 3.0.4 patch release.

**Comment by colin-lamed** [comment]
This sounds great for us. We certainly don't have anywhere near 22 parameters.

We'd be happy to wait for the 3.0.4 release - since the trait is just used in tests we can stick with 3.0.2 for now.
